clang/test/Sema/attr-model.cpp

```diff
@@ -1,64 +1,40 @@
-// RUN: %clang_cc1 -triple aarch64 -verify=expected,aarch64 -fsyntax-only %s
+// RUN: %clang_cc1 -triple aarch64 -verify=expected,unsupported -fsyntax-only %s
 // RUN: %clang_cc1 -triple loongarch64 -verify=expected,loongarch64 -fsyntax-only %s
-// RUN: %clang_cc1 -triple mips64 -verify=expected,mips64 -fsyntax-only %s
-// RUN: %clang_cc1 -triple powerpc64 -verify=expected,powerpc64 -fsyntax-only %s
-// RUN: %clang_cc1 -triple riscv64 -verify=expected,riscv64 -fsyntax-only %s
+// RUN: %clang_cc1 -triple mips64 -verify=expected,unsupported -fsyntax-only %s
+// RUN: %clang_cc1 -triple powerpc64 -verify=expected,unsupported -fsyntax-only %s
+// RUN: %clang_cc1 -triple riscv64 -verify=expected,unsupported -fsyntax-only %s
 // RUN: %clang_cc1 -triple x86_64 -verify=expected,x86_64 -fsyntax-only %s
+// RUN: %clang_cc1 -triple nvptx64-unknown-cuda -fcuda-is-device -x cuda -verify=expected,unsupported -fsyntax-only %s
```

Comment:
If the test passes this way (I.e. you do see complaints about the attributes) that means that the change *will* break CUDA.

Considering that we will see these attributes in the host code and that NVPTX itself can't do anything useful with them, NVPTX compilation should continue to work and ignore those attributes.

Reply:
The attribute *is* technically is ignored, with a warning, but the warning is likely non-actionable as it would be issues for a perfectly correct host code which just happens to be seen by a GPU compilation.

I think the correct way to handle it would be to issue a deferred diagnostic which will fire only if we end up code-gen-ing something that uses the attribute.

@yxsamliu Sam, does that make sense?

Reply:
I see, although it seems pretty annoying to have to special case NVPTX compilations. Is there a previous example of having done something like this, or should I just ad-hoc skip the Sema code for NVPTX compilations?

Reply:
> I think the correct way to handle it would be to issue a deferred diagnostic which will fire only if we end up code-gen-ing something that uses the attribute.

is it common to define a global in a TU that's shared between the host and device?

Reply:
Yes. global scope is implicitly considered to belong to host. Data that's supposed to reside on the device must be explicitly annotated with appropriate attributes.

Reply:
I'm tempted to keep things as is since it seems like this is not the only issue that can come up with nvptx compiles. For example, the `thread_local` global in the variable has the same issue where nvptx doesn't support thread local globals, which clang complains about, and I don't really see a difference between `__attribute__((model()))` and `thread_local` in that they can both be used on arbitrary globals.

Do you think it's reasonable to postpone this suggestion until people actually hit it? For example, if people aren't hitting the `thread_local` issue, then perhaps people won't hit a warning with this attribute.

Reply:
Up to you.

Reply:
> The attribute _is_ technically is ignored, with a warning, but the warning is likely non-actionable as it would be issues for a perfectly correct host code which just happens to be seen by a GPU compilation.
> 
> I think the correct way to handle it would be to issue a deferred diagnostic which will fire only if we end up code-gen-ing something that uses the attribute.
> 
> @yxsamliu Sam, does that make sense?

My understanding is that deferred diagnostics only works in a function, but this diagnostic happens outside functions.